pkg/snapshot/restorer/restorer.go

```diff
@@ -317,28 +322,214 @@ func startEmbeddedEtcd(ro RestoreOptions) (*embed.Etcd, error) {
+
+			eventsData, err := getEventsDataFromDeltaSnapshot(r.store, fetcherInfo.Snapshot)
+			if err != nil {
+				errCh <- fmt.Errorf("failed to read events data from delta snapshot %s : %v", fetcherInfo.Snapshot.SnapName, err)
+				applierInfoCh <- applierInfo{SnapIndex: -1} // cannot use close(ch) as concurrent fetchSnaps routines might try to send on channel, causing a panic
+				return
+			}
+
+			eventsFilePath, err := persistDeltaSnapshot(eventsData)
+			if err != nil {
+				errCh <- fmt.Errorf("failed to persist events data for delta snapshot %s : %v", fetcherInfo.Snapshot.SnapName, err)
+				applierInfoCh <- applierInfo{SnapIndex: -1}
+				return
+			}
+
+			eventsData = nil // ensures that underlying memory gets garbage-collected
+
+			snapLocationsCh <- eventsFilePath // used for cleanup later
+
+			applierInfo := applierInfo{
+				EventsFilePath: eventsFilePath,
+				SnapIndex:      fetcherInfo.SnapIndex,
+			}
+			applierInfoCh <- applierInfo
+		}
+	}
+}
+
+// apply applies delta snapshot events to the embedded etcd sequentially, in the right order of snapshots, regardless of the order in which they were fetched.
+func (r *Restorer) applySnaps(client *clientv3.Client, remainingSnaps snapstore.SnapList, applierInfoCh <-chan applierInfo, errCh chan<- error, stopCh <-chan bool, wg *sync.WaitGroup) {
+	defer func() {
+		wg.Done()
+	}()
+
+	wg.Add(1)
+
+	eventsList := make([][]event, len(remainingSnaps))
+	snapFetchStatus := make([]bool, len(remainingSnaps))
+	for i := 0; i < len(snapFetchStatus); i++ {
+		snapFetchStatus[i] = false
+	}
+	nextSnapIndexToApply := 0
+
+	for i := 0; i < len(remainingSnaps); i++ {
+		select {
+		case _, more := <-stopCh:
+			if !more {
+				return
+			}
+		case applierInfo := <-applierInfoCh:
+			if applierInfo.SnapIndex == -1 {
+				return
+			}
+			filePath := applierInfo.EventsFilePath
+			fetchedSnapIndex := applierInfo.SnapIndex
+			snapName := remainingSnaps[fetchedSnapIndex].SnapName
+
+			eventsData, err := ioutil.ReadFile(filePath)
+			if err != nil {
+				errCh <- fmt.Errorf("failed to read events data from file for delta snapshot %s : %v", snapName, err)
+				return
+			}
+
+			if err = os.Remove(filePath); err != nil {
+				r.logger.Warnf("Unable to remove file: %s; err: %v", filePath, err)
+			}
+
+			events := []event{}
+			if err = json.Unmarshal(eventsData, &events); err != nil {
+				errCh <- fmt.Errorf("failed to read events from events data for delta snapshot %s : %v", snapName, err)
+				return
+			}
+
+			eventsList[fetchedSnapIndex] = events
+			snapFetchStatus[fetchedSnapIndex] = true
+
+			if fetchedSnapIndex < nextSnapIndexToApply {
+				errCh <- fmt.Errorf("snap index mismatch for delta snapshot %d; expected snap index to be atleast %d", fetchedSnapIndex, nextSnapIndexToApply)
+				return
+			}
+			if fetchedSnapIndex == nextSnapIndexToApply {
```

Comment:
What if `fetchedSnapIndex > nextSnapIndexToApply`?

Reply:
It will just continue to the next iteration. The logic is: when `fetchedSnapIndex` == `nextSnapIndexToApply`, it applies that snap as well as any existing subsequent snaps till an unfetched snap is encountered, and it moves to the next iteration. This way, all snaps are applied in the right order.